packages/flutter/lib/src/widgets/_window.dart

```diff
@@ -0,0 +1,542 @@
+  set rootView(FlutterView view) {
+    _view = view;
+  }
+
+  /// The type of the window.
+  ///
+  /// {@macro flutter.widgets.windowing.experimental}
+  @internal
+  WindowType get type;
+
+  /// The current size of the window.
+  ///
+  /// This might differ from the requested size.
+  ///
+  ///
+  /// This might also differ from the actual size of the window if the window has
+  /// decorations such as title bar, borders, etc.
+  ///
+  /// {@macro flutter.widgets.windowing.experimental}
+  @internal
+  Size get contentSize;
+
+  /// Destroys this window.
+  ///
+  /// It is permissible to call this method multiple times.
+  ///
+  /// {@macro flutter.widgets.windowing.experimental}
+  @internal
+  void destroy();
+
+  /// The root view associated to this window, which is unique to each window.
+  ///
+  /// {@macro flutter.widgets.windowing.experimental}
+  @internal
+  FlutterView get rootView => _view;
+  late final FlutterView _view;
+}
+
+/// Delegate class for regular window controller.
+///
+/// {@macro flutter.widgets.windowing.experimental}
+@internal
+mixin class RegularWindowControllerDelegate {
+  /// Invoked when user attempts to close the window. Default implementation
+  /// destroys the window. Subclass can override the behavior to delay
+  /// or prevent the window from closing.
+  ///
+  /// {@macro flutter.widgets.windowing.experimental}
+  @internal
+  void onWindowCloseRequested(RegularWindowController controller) {
+    if (!isWindowingEnabled) {
+      throw UnsupportedError(_windowingDisabledErrorMessage);
+    }
+
+    controller.destroy();
+  }
+
+  /// Invoked when the window is closed.
+  ///
+  /// Default implementation exits the application if this was
+  /// the last top-level window.
+  ///
+  /// {@macro flutter.widgets.windowing.experimental}
+  @internal
+  void onWindowDestroyed() {
+    if (!isWindowingEnabled) {
+      throw UnsupportedError(_windowingDisabledErrorMessage);
+    }
+
+    final WindowingOwner owner = WidgetsBinding.instance.windowingOwner;
+    if (!owner.hasTopLevelWindows()) {
+      // TODO(mattkae): close the application if this is the last window
+      // via ServicesBinding.instance.exitApplication(AppExitType.cancelable);
+    }
+  }
+}
+
+/// A controller for a regular window.
+///
+/// A regular window is a traditional window that can be resized, minimized,
+/// maximized, and closed. Upon construction, the window is created for the
+/// platform with the provided properties.
+///
+/// This class does not interact with the widget tree. Instead, it is typically
+/// provided to the [RegularWindow] widget, who does the work of rendering the
+/// content inside of this window.
+///
+/// An example usage might look like:
+/// ```dart
+/// final RegularWindowController controller = RegularWindowController(
+///   contentSize: const WindowSizing(
+///     preferredSize: Size(800, 600),
+///     preferredConstraints: BoxConstraints(minWidth: 640, minHeight: 480),
+///   ),
```

Comment:
Just curious, what's the benefit of the `WindowSizing` type? In the https://flutter.dev/go/multi-window-api design doc, we instead made `RegularWindowController`'s constructor and the `updateContentSize` method accept a size & constraints directly. Are we planning on adding a bunch more stuff to `WindowSizing` in the future?

For example:

```dart
final RegularWindowController controller = RegularWindowController(
  preferredSize: Size(800, 600),
  preferredConstraints: BoxConstraints(minWidth: 640, minHeight: 480),
  title: "Example Window",
);
```

Reply:
In that future world, what are all the members that would be on `WindowSizing`?

Reply:
I do think all these should be considered together. And while the platform could well reason about conflicting values (e.g. prioritize size constraints over preferred size), we have an opportunity not to, and error out on invalid values.

Reply:
Sounds good, that should probably be checked with an assertion so that it only shows up in debug mode.